src/hawkbit-client.c

```diff
@@ -265,40 +267,42 @@ static size_t curl_write_cb(void *content, size_t size, size_t nmemb, void *data
 }
 
 /**
- * @brief Make REST request.
+ * @brief Perform REST request with JSON data, expecting response JSON data.
  *
- * @param[in]  method             HTTP Method ex. GET
+ * @param[in]  method             HTTP Method, e.g. GET
  * @param[in]  url                URL used in HTTP REST request
- * @param[in]  jsonRequestBody    REST request body. If NULL, no body is sent.
- * @param[out] jsonResponseParser REST response
+ * @param[in]  jsonRequestBody    REST request body. If NULL, no body is sent
+ * @param[out] jsonResponseParser REST response. If NULL, response is not parsed
  * @param[out] error              Error
  * @return TRUE if request and response parser (if given) suceeded, FALSE otherwise (error set).
  */
-static gboolean rest_request(enum HTTPMethod method, const gchar* url, JsonBuilder* jsonRequestBody, JsonParser** jsonResponseParser, GError** error)
+static gboolean rest_request(enum HTTPMethod method, const gchar *url,
+                             JsonBuilder *jsonRequestBody, JsonParser **jsonResponseParser,
+                             GError **error)
 {
-        gchar *postdata = NULL;
-        g_autofree gchar *token = NULL;
-        RestPayload *fetch_buffer = g_new0(RestPayload, 1);
-        glong http_code = 0;
+        g_autofree gchar *postdata = NULL, *token = NULL;
+        g_autoptr(RestPayload) fetch_buffer = NULL;
         struct curl_slist *headers = NULL;
-        CURL *curl = curl_easy_init();
-        gboolean ret = TRUE;
+        g_autoptr(CURL) curl = NULL;
+        glong http_code = 0;
+        CURLcode res;
+
+        g_return_val_if_fail(url, FALSE);
+        g_return_val_if_fail(jsonResponseParser == NULL || *jsonResponseParser == NULL, FALSE);
+        g_return_val_if_fail(error == NULL || *error == NULL, FALSE);
+
+        curl = curl_easy_init();
         if (!curl) {
                 g_set_error(error, RHU_HAWKBIT_CLIENT_CURL_ERROR, -1, "curl init failed");
-                ret = FALSE;
-                goto clean;
+                return FALSE;
```

Comment:
This actually reverts flow changes made in the commit before, maybe doing the autocleanup support earlier will a bunch of unnecessary changes?

Reply:
Yep, changing the return type and refactoring of `rest_request()` got mixed up, so the whole intermediate flow changes can be dropped. I'll do that.